lib/charms/operator_libs_linux/v0/apt.py

```diff
@@ -1315,47 +1409,293 @@ def _parse(line: str, filename: str) -> DebianRepository:
             raise InvalidSourceError("An invalid sources line was found in %s!", filename)
 
     def add(self, repo: DebianRepository, default_filename: Optional[bool] = False) -> None:
-        """Add a new repository to the system.
-
-        Args:
-          repo: a `DebianRepository` object
-          default_filename: an (Optional) filename if the default is not desirable
+        """Add a new repository to the system. May have destructive or surprising behaviour.
+
+        Will write in the deb822 format if:
+            1. repo.filename has the '.sources' extension
+            2. repo.filename is unset (""), but /etc/apt/sources.list.d/ubuntu.sources exists
+        Otherwise the one-per-line style will be used.
+
+        WARNING: the default_filename keyword argument is provided for backwards compatibility
+        only. It is not used, and was not used in the previous revision of this library.
+
+        WARNING: in the one-per-line format case, will mutate repo.options to add the 'signed-by'
+        key if both options and gpg_file are truthy (for example if a gpg_key_filename and
+        non-empty options were provided at DebianRepository initialisation time). If options were
+        not provided or are empty, but a gpg_key is available, will silently fail to include it.
+
+        WARNING: if repo.filename is falsey, the new filename is calculated only from the repo's
+        uri and release, but repos are assumed to be uniquely identified by the combination of
+        repotype, uri, and release. Adding two repos with differing repotypes but identical uris
+        and releases will result in the second repo clobbering the file written by the first.
+        In this case, repo.filename must be set appropriately to avoid data loss.
+
+        WARNING: if repo.filename is truthy, and that file exists, this method will clobber that
+        file with a single entry for the repo being added. Set it to a falsey value to have a new
+        filename derived from its uri and release (which will also clobber any existing file), or
+        set the filename that you want to write to -- or construct a new DebianRepository object
+        with the filename you want to write to.
+        For example: repo.filename = my_filename
+        For example: DebianRepository(uri=repo.uri, filename=my_filename, ...)
```

Comment:
I don't love this behaviour. It seems especially bad that any `DebianRepository` `repo` created via a `RepositoryMapping` `mapping` will clobber the entire file it was loaded from with a single entry if you call `mapping.add(repo)`.